text/0000-network-spec.md

```diff
@@ -0,0 +1,160 @@
+- Proposal Name: Network specification
+- Start Date: 2019-08-26
+- NEP PR: [nearprotocol/neps#0000](https://github.com/nearprotocol/neps/pull/0000)
+
+# Summary
+[summary]: #summary
+
+Specification of the networking layer that peers use to communicate.
+Includes implementation details of peer discovery, handshake and reputation.
+Does not include the format of the actual blockchain protocol.
+
+# Motivation
+[motivation]: #motivation
+
+To provide actually secure routing, especially among the validators, we postulate that account information and cryptographic signatures backed by stake are required.
+This network specification describes all of the protocols to connect, handshake, discover peers and maintain connections and an integration with the blockchain itself.
+
+This document should serve as reference for all the clients to implement networking layer.
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+## System description
+
+Our system is design on the random sampling of peers and cryptohraphic routing.
+Each node has ID that is ED25519 public key derived from secret key known only to this node.
+Additionally, active participants, such as block producers and other nodes that want to link their account information to the peer information have a string `account_id` and appropriate `public_key` which is associated with this account. Accounts maintained by blockchain itself, and can not be forged without breaking the blockchain itself.
+
+### Storing network information
+
+A client stores information about other nodes in three data structures. `known_peers` a long-term database, which is stored on disk and persists between reboots. And two short-term databases `active_peers` and `routing` that start empty.
+
+`known_peers`. The `known_peers` is stored on disk and contains information about each node that client has ever connected or recevied gossip about. There is no limit to the size of the `known_peers`.
+Each entry of `known_peers` is tuple of `<node ID, IP address, TCP port, last_seen, status>`. Where `last_seen` is a timestamp when we received last message from this peer. If this was received from gossip, `last_seen` is 0. Status contains if this peer is currently unknown, connected, disconnected or banned.
```

Comment:
Is this IPv4 or IPv6?

IPv6 would invalidate any attempt at IP bans, since you can have millions of addresses for a single computer with little effort

IPv4 is a smaller space and is more prone to directly identifying the peer, which is bad.

If known peers are tuples of `node ID` and `IP address` does that mean that a client that moves from one IP to another would come out as a brand new client, even though it has the same `node ID` as a previous peer?

Reply:
We are trying to address this issue in the new network design using a reputation system.